README.md
# DCPU-16 Emulator and Assembler #

Based on the specification [here](http://0x10c.com/doc/dcpu-16.txt).

Includes an extra instruction HLT which stops execution of the current program
(otherwise it will run forever unless you tell it to only execute a certain
number of operations).

## Emulator ##

`perl -Ilib bin/dcpu16-execute [--iterations <n>] [--dump <outfile>] <binary>`

Executes a DCPU-16 binary file. Runs until a HLT instruction is seen, unless
`--iterations` is specified, in which case it runs for that many clock cycles.
Produces a memory and register dump at the end in the file `./dcpu16.dump` (or
whatever file you specify with the `--dump` option).

## Assembler ##

`perl -Ilib bin/dcpu16-asm [--output <outfile>] <asm_file>`

Creates a DCPU-16 binary from an assembler file. Assembler syntax is so far
limited to the examples that are given in the specification. Output is written
to `./a.out`, or whatever filename is given in the `--output` option.


Changes
Revision history for Games-Emulation-DCPU16

{{$NEXT}}
      - Initial release


examples/spec.asm
; Try some basic stuff
              SET A, 0x30              ; 7c01 0030
              SET [0x1000], 0x20       ; 7de1 1000 0020
              SUB A, [0x1000]          ; 7803 1000
              IFN A, 0x10              ; c00d 
                 SET PC, crash         ; 7dc1 001a [*]

; Do a loopy thing
              SET I, 10                ; a861
              SET A, 0x2000            ; 7c01 2000
:loop         SET [0x2000+I], [A]      ; 2161 2000
              SUB I, 1                 ; 8463
              IFN I, 0                 ; 806d
                 SET PC, loop          ; 7dc1 000d [*]

; Call a subroutine
              SET X, 0x4               ; 9031
              JSR testsub              ; 7c10 0018 [*]
              SET PC, crash            ; 7dc1 001a [*]

:testsub      SHL X, 4                 ; 9037
              SET PC, POP              ; 61c1

; Hang forever. X should now be 0x40 if everything went right.
:crash        SET PC, crash            ; 7dc1 001a [*]

; [*]: Note that these can be one word shorter and one cycle faster by using the short form (0x00-0x1f) of literals,
;      but my assembler doesn't support short form labels yet.     


lib/Games/Emulation/DCPU16/Util.pm
package Games::Emulation::DCPU16::Util;
use strict;
use warnings;

use Exporter 'import';

our @EXPORT_OK = ('format_registers', 'format_memory');

sub format_registers {
    my ($registers, $sp, $pc, $o) = @_;

    return join("\n",
        sprintf("A:  0x%04x    X:  0x%04x    I:  0x%04x",
                $registers->[0], $registers->[3], $registers->[6]),
        sprintf("B:  0x%04x    Y:  0x%04x    J:  0x%04x",
                $registers->[1], $registers->[4], $registers->[7]),
        sprintf("C:  0x%04x    Z:  0x%04x",
                $registers->[2], $registers->[5]),
        "",
        sprintf("SP: 0x%04x    PC: 0x%04x    O:  0x%04x", $sp, $pc, $o),
        "",
    );
}

sub format_memory {
    my ($memory) = @_;

    my $out = '';
    my $eliding = 0;
    for my $addr (0..int(@$memory / 8)) {
        $addr *= 8;
        my @values = @{ $memory }[$addr..($addr + 7)];
        if (grep { $_ } @values) {
            $out .= sprintf("%04x:" . (" %04x" x 8), $addr, @values) . "\n";
            $eliding = 0;
        }
        else {
            if (!$eliding) {
                $out .= "...\n";
                $eliding = 1;
            }
        }
    }

    return $out;
}

1;


lib/Games/Emulation/DCPU16/Assembler.pm
package Games::Emulation::DCPU16::Assembler;
use strict;
use warnings;

sub new {
    my $class = shift;
    bless {
        bytes      => '',
        line       => 0,

        labels     => {},
        unresolved => {},

        basic_ops => {
            SET => 0x01,
            ADD => 0x02,
            SUB => 0x03,
            MUL => 0x04,
            DIV => 0x05,
            MOD => 0x06,
            SHL => 0x07,
            SHR => 0x08,
            AND => 0x09,
            BOR => 0x0a,
            XOR => 0x0b,
            IFE => 0x0c,
            IFN => 0x0d,
            IFG => 0x0e,
            IFB => 0x0f,
        },
        non_basic_ops => {
            JSR => 0x01,
            HLT => 0x3f,
        },
        registers => {
            A => 0x00,
            B => 0x01,
            C => 0x02,
            X => 0x03,
            Y => 0x04,
            Z => 0x05,
            I => 0x06,
            J => 0x07,
        },
    }, $class;
}

sub bytes { shift->{bytes} }

sub assemble {
    my $self = shift;
    my ($script) = @_;

    for my $line (split /\n/, $script) {
        $self->parse_line($line);
    }

    $self->resolve_references;

    return $self->{bytes};
}

sub parse_line {
    my $self = shift;
    my ($line) = @_;

    $self->{line}++;

    my $clean_line = $self->_clean_line($line);
    return unless length($clean_line);

    my ($label, $op, $a, $b) = $clean_line =~ m!
        ^              \s*
        (?::(\w+)      \s+)?
        ([A-Z]{3})     \s+
        ([^,\s]+) (?:, \s+
        ([^,\s]+))?    \s*
        $
    !x;

    die "Couldn't parse \"$line\" (line $self->{line})"
        unless defined $op;

    $self->{labels}{$label} = length($self->{bytes}) / 2
        if defined $label;

    $op = uc($op);
    if (my $basic_opcode = $self->{basic_ops}{$op}) {
        die "$op requires two values (line $self->{line})"
            unless defined($b);

        my ($val1, $next_word1, $label1) = $self->_parse_value($a);
        my ($val2, $next_word2, $label2) = $self->_parse_value($b);

        $basic_opcode |= $val1 << 4;
        $basic_opcode |= $val2 << 10;

        $self->{unresolved}{length($self->{bytes}) / 2} = [ $label1, $label2 ]
            if defined($label1) || defined($label2);

        $self->{bytes} .= pack("S>", $basic_opcode);
        $self->{bytes} .= pack("S>", $next_word1) if defined $next_word1;
        $self->{bytes} .= pack("S>", $next_word2) if defined $next_word2;
    }
    elsif (my $non_basic_opcode = $self->{non_basic_ops}{$op}) {
        my ($val, $next_word, $label) = $self->_parse_value($a);

        $non_basic_opcode <<= 4;
        $non_basic_opcode |= $val << 10;

        $self->{unresolved}{length($self->{bytes}) / 2} = [ $label ]
            if defined($label);

        $self->{bytes} .= pack("S>", $non_basic_opcode);
        $self->{bytes} .= pack("S>", $next_word) if defined $next_word;
    }
    else {
        die "Invalid op: $op (line $self->{line})";
    }
}

sub resolve_references {
    my $self = shift;

    for my $pos (reverse sort { $a <=> $b } keys %{ $self->{unresolved} }) {
        my @labels = grep { defined } @{ delete $self->{unresolved}{$pos} };
        next unless @labels;

        my $offset = 2;
        for my $label (@labels) {
            die "Unknown label $label (during resolution)"
                unless exists $self->{labels}{$label};

            # XXX collapse small integers
            substr(
                $self->{bytes},
                $pos * 2 + $offset,
                2,
                pack("S>", $self->{labels}{$label})
            );
            $offset += 2;
        }
    }
}

sub _clean_line {
    my $self = shift;
    my ($line) = @_;

    $line =~ s/;.*//;
    $line =~ s/^\s*|\s*$//;
    $line =~ s/\s+/ /g;

    return $line;
}

sub _parse_value {
    my $self = shift;
    my ($value) = @_;

    my $reg = qr/[ABCXYZIJ]/;
    my $num = qr/(?:0[xb])?[0-9]+/;

    if ($value =~ /^($reg)$/) {
        return ($self->{registers}{$1});
    }
    elsif ($value =~ /^\[\s*($reg)\s*\]$/) {
        return (0x08 + $self->{registers}{$1});
    }
    elsif ($value =~ /^\[\s*($num)\s*\+\s*($reg)\s*\]$/) {
        return (0x10 + $self->{registers}{$2}, $self->_parse_num($1));
    }
    elsif ($value eq 'POP'  || $value =~ /^\[\s*SP\+\+\s*\]$/) {
        return (0x18);
    }
    elsif ($value eq 'PEEK' || $value =~ /^\[\s*SP\s*\]$/) {
        return (0x19);
    }
    elsif ($value eq 'PUSH' || $value =~ /^\[\s*--SP\s*\]$/) {
        return (0x1a);
    }
    elsif ($value eq 'SP') {
        return (0x1b);
    }
    elsif ($value eq 'PC') {
        return (0x1c);
    }
    elsif ($value eq 'O') {
        return (0x1d);
    }
    elsif ($value =~ /^\[\s*($num)\s*\]$/) {
        return (0x1e, $self->_parse_num($1));
    }
    elsif ($value =~ /^($num)$/) {
        my $num = $self->_parse_num($1);
        if ($num < 0x20) {
            return 0x20 + $num;
        }
        else {
            return (0x1f, $self->_parse_num($1));
        }
    }
    elsif ($value =~ /\w+/) {
        return (0x1f, 0x00, $value);
    }
    else {
        die "Can't parse value \"$value\" (line $self->{line})";
    }
}

sub _parse_num {
    my $self = shift;
    my ($num) = @_;

    die "Invalid number $num (line $self->{line})"
        unless $num =~ /^(?:0[xb])?[0-9]+/;

    my $decimal = $num;
    $decimal = oct($num) if $num =~ /^0/;

    die "Number $decimal too large (line $self->{line})"
        if $decimal >= 2**16;

    return $decimal;
}

1;


lib/Games/Emulation/DCPU16.pm
package Games::Emulation::DCPU16;
use strict;
use warnings;

use integer;

use constant {
    STATE_NEW_OP     => 0,
    STATE_READ_ARG_1 => 1,
    STATE_READ_ARG_2 => 2,
    STATE_OP_EXECUTE => 3,
    TIME_TAKEN_NONE  => 0,
    TIME_TAKEN_DELAY => 1,
    TIME_TAKEN_WORK  => 2,
};

sub new {
    my $class = shift;

    bless {
        memory           => [(0x0000) x 0x10000],
        registers        => [(0x0000) x 8],
        PC               => 0x0000,
        SP               => 0x0000,
        O                => 0x0000,

        clock            => 0,

        halt             => undef,
        delay            => 0,
        time_taken       => TIME_TAKEN_NONE,
        state            => STATE_NEW_OP,
        has_delayed      => undef,

        current_op       => undef,
        lvalue           => undef,
        value1           => undef,
        value2           => undef,
        next_word        => undef,
        pc_offset        => 0,
        jumped           => undef,

        basic_opcode_map => [
            undef,
            '_op_SET',
            '_op_ADD',
            '_op_SUB',
            '_op_MUL',
            '_op_DIV',
            '_op_MOD',
            '_op_SHL',
            '_op_SHR',
            '_op_AND',
            '_op_BOR',
            '_op_XOR',
            '_op_IFE',
            '_op_IFN',
            '_op_IFG',
            '_op_IFB',
        ],
        non_basic_opcode_map => [
            undef,
            '_op_JSR',
            (undef) x (0x3f - 0x02 - 1),
            '_op_HLT', # XXX extension
        ],
    }, $class;
}

sub memory    { shift->{memory}    }
sub registers { shift->{registers} }
sub PC        { shift->{PC}        }
sub SP        { shift->{SP}        }
sub O         { shift->{O}         }
sub clock     { shift->{clock}     }

sub load {
    my $self = shift;
    my ($bytecode) = @_;

    my $idx = 0;
    while (my $word = substr($bytecode, 0, 2, '')) {
        $self->{memory}[$idx++] = ord($word) * 2**8 + ord(substr($word, 1, 1));
    }
}

sub run {
    my $self = shift;

    $self->step until $self->{halt};
}

sub step {
    my $self = shift;

    $self->{clock}++;

    if ($self->{delay}) {
        $self->{delay}--;
        return;
    }

    $self->{time_taken} = TIME_TAKEN_NONE;
    while (1) {
        my $state = $self->{state};
        if ($state == STATE_NEW_OP) {
            $self->{state} = $self->_parse_op($self->{memory}[$self->{PC}]);
            $self->{pc_offset} = 1;
            undef $self->{jumped};
        }
        elsif ($state == STATE_READ_ARG_1) {
            $self->{state} = $self->_parse_value($self->{value1})
        }
        elsif ($state == STATE_READ_ARG_2) {
            $self->{state} = $self->_parse_value($self->{value2})
        }
        elsif ($state == STATE_OP_EXECUTE) {
            $self->{state} = $self->_execute_current_op;
        }
        else {
            die "Invalid state";
        }

        last if $self->{time_taken} != TIME_TAKEN_NONE;
    }
}

# XXX this duplicates a bit from _parse_value
sub _op_length {
    my $self = shift;

    my $length = 1;
    $self->_parse_op($self->{memory}[$self->{PC} + $self->{pc_offset}]);

    for my $value ($self->{value1}, $self->{value2}) {
        next unless defined $value;
        $length++ if $value >= 0x10 && $value < 0x18;
        $length++ if $value >= 0x1e && $value < 0x20;
    }

    return $length;
}

sub _parse_op {
    my $self = shift;
    my ($opcode) = @_;

    my $basic_op = $opcode & 0x0f;

    if ($basic_op) {
        $self->{value1}     = ($opcode >> 4)  & 0x3f;
        $self->{value2}     = ($opcode >> 10) & 0x3f;
        $self->{current_op} = $self->{basic_opcode_map}[$basic_op];
    }
    else {
        my $non_basic_op = ($opcode >> 4) & 0x3f;

        $self->{value1}     = ($opcode >> 10) & 0x3f;
        $self->{value2}     = undef;
        $self->{current_op} = $self->{non_basic_opcode_map}[$non_basic_op];
    }

    die "Illegal opcode" unless $self->{current_op};

    return STATE_READ_ARG_1;
}

sub _parse_value {
    my $self = shift;
    my ($value) = @_;

    my $state = $self->{state};
    my $key = $state == STATE_READ_ARG_1 ? 'value1' : 'value2';

    if ($value < 0x08) {
        $self->{$key} = $self->{registers}[$value];
        $self->{lvalue} = \$self->{registers}[$value]
            if $state == STATE_READ_ARG_1;
    }
    elsif ($value < 0x10) {
        my $addr = $self->{registers}[$value & 0x07];
        $self->{$key} = $self->{memory}[$addr];
    }
    elsif ($value < 0x18) {
        $self->_next_word;
        return $state if $self->{time_taken} != TIME_TAKEN_WORK;

        my $addr = $self->{registers}[$value & 0x07] + $self->{next_word};
        $self->{$key} = $self->{memory}[$addr];
        $self->{lvalue} = \$self->{memory}[$addr]
            if $state == STATE_READ_ARG_1;
    }
    elsif ($value > 0x1f) {
        die "Illegal value" unless $value <= 0x3f;
        $self->{$key} = $value - 0x20;
    }
    elsif ($value == 0x18) {
        $self->{$key} = $self->{memory}[($self->{SP}++ & 0xffff)];
        $self->{SP} &= 0xffff;
    }
    elsif ($value == 0x19) {
        $self->{$key} = $self->{memory}[$self->{SP}];
    }
    elsif ($value == 0x1a) {
        $self->{$key} = $self->{memory}[(--$self->{SP} & 0xffff)];
        $self->{SP} &= 0xffff;
    }
    elsif ($value == 0x1b) {
        $self->{$key} = $self->{SP};
        $self->{lvalue} = \$self->{SP}
            if $state == STATE_READ_ARG_1;
    }
    elsif ($value == 0x1c) {
        $self->{$key} = $self->{PC};
        $self->{lvalue} = \$self->{PC}
            if $state == STATE_READ_ARG_1;
    }
    elsif ($value == 0x1d) {
        $self->{$key} = $self->{O};
    }
    elsif ($value == 0x1e) {
        $self->_next_word;
        return $state if $self->{time_taken} != TIME_TAKEN_WORK;

        my $addr = $self->{next_word};
        $self->{$key} = $self->{memory}[$addr];
        $self->{lvalue} = \$self->{memory}[$addr]
            if $state == STATE_READ_ARG_1;
    }
    elsif ($value == 0x1f) {
        $self->_next_word;
        return $state if $self->{time_taken} != TIME_TAKEN_WORK;

        $self->{$key} = $self->{next_word};
    }

    return $state == STATE_READ_ARG_2 || !defined($self->{value2})
        ? STATE_OP_EXECUTE
        : STATE_READ_ARG_2;
}

sub _execute_current_op {
    my $self = shift;

    my $op_meth = $self->{current_op};
    $self->$op_meth($self->{value1}, $self->{value2});

    return $self->{state} if $self->{time_taken} != TIME_TAKEN_WORK;

    $self->{PC} += $self->{pc_offset}
        unless $self->{jumped};
    $self->{lvalue} = undef;

    return STATE_NEW_OP;
}

sub _next_word {
    my $self = shift;

    undef $self->{next_word};

    return 1 if $self->_delay(1);

    $self->{next_word} = $self->{memory}[$self->{PC} + $self->{pc_offset}];
    $self->{pc_offset}++;

    return;
}

sub _delay {
    my $self = shift;
    my ($delay) = @_;

    return unless $delay;

    $delay--;

    if (!$delay || $self->{has_delayed}) {
        $self->{time_taken} = TIME_TAKEN_WORK;
        undef $self->{has_delayed};
        return;
    }
    else {
        $self->{time_taken} = TIME_TAKEN_DELAY;
        $self->{delay} = $delay - 1;
        $self->{has_delayed} = 1;
        return 1;
    }
}

sub _op_SET {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(1);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $self->{jumped} = 1
        if $lvalue == \$self->{PC};

    ${ $self->{lvalue} } = $b;
}

sub _op_ADD {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(2);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a + $b;
    $self->{O} = $$lvalue >> 16;
    $$lvalue &= 0xffff;
}

sub _op_SUB {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(2);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a - $b;
    $self->{O} = $$lvalue >> 16;
    $$lvalue &= 0xffff;
}

sub _op_MUL {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(2);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a * $b;
    $self->{O} = $$lvalue >> 16;
    $$lvalue &= 0xffff;
}

sub _op_DIV {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(3);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    if ($b == 0) {
        $$lvalue = 0;
        $self->{O} = 0;
    }
    else {
        $$lvalue = $a / $b;
        $self->{O} = (($a << 16) / $b) & 0xffff;
        $$lvalue &= 0xffff;
    }
}

sub _op_MOD {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(3);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    if ($b == 0) {
        $$lvalue = 0;
    }
    else {
        $$lvalue = ($a % $b) & 0xffff;
    }
}

sub _op_SHL {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(2);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a << $b;
    $self->{O} = $$lvalue >> 16;
    $$lvalue &= 0xffff;
}

sub _op_SHR {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(2);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a >> $b;
    $self->{O} = (($a << 16) >> $b) & 0xffff;
    $$lvalue &= 0xffff;
}

sub _op_AND {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(1);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a & $b;
}

sub _op_BOR {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(1);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a | $b;
}

sub _op_XOR {
    my $self = shift;
    my ($a, $b) = @_;

    return if $self->_delay(1);

    my $lvalue = $self->{lvalue};
    return unless $lvalue;

    $$lvalue = $a ^ $b;
}

sub _op_IFE {
    my $self = shift;
    my ($a, $b) = @_;

    if ($a == $b) {
        return if $self->_delay(2);
    }
    else {
        return if $self->_delay(3);

        $self->{pc_offset} += $self->_op_length;
    }
}

sub _op_IFN {
    my $self = shift;
    my ($a, $b) = @_;

    if ($a != $b) {
        return if $self->_delay(2);
    }
    else {
        return if $self->_delay(3);

        $self->{pc_offset} += $self->_op_length;
    }
}

sub _op_IFG {
    my $self = shift;
    my ($a, $b) = @_;

    if ($a > $b) {
        return if $self->_delay(2);
    }
    else {
        return if $self->_delay(3);

        $self->{pc_offset} += $self->_op_length;
    }
}

sub _op_IFB {
    my $self = shift;
    my ($a, $b) = @_;

    if ($a & $b) {
        return if $self->_delay(2);
    }
    else {
        return if $self->_delay(3);

        $self->{pc_offset} += $self->_op_length;
    }
}

sub _op_JSR {
    my $self = shift;
    my ($a) = @_;

    return if $self->_delay(2);

    $self->{memory}[(--$self->{SP} & 0xffff)] = $self->{PC} + $self->{pc_offset};
    $self->{SP} &= 0xffff;
    $self->{PC} = $a;
    $self->{jumped} = 1;
}

sub _op_HLT {
    my $self = shift;
    my ($a) = @_;

    $self->{halt} = 1;
}

=for notes

what happens when an invalid op is read?

=cut

1;
